### PR TITLE
Added thread_requests parameter to import_pbp_data and import_weekly …

### PR DESCRIPTION
**Pull request by bendominguez0111**
…data functions.

Added optional parameter to import_pbp_data and import_weekly_data to use threading to speed up requests for play by play data and weekly data. I also tested async and multiprocessing as well but threading posted the best results. Depending on connection, it sped up the speed at which PBP data from 1999 to 2022 was loaded by 25-50%. 

Also added associated tests. I added a __init__.py file to the tests folder, because it was the only way I could run pytest (although I may have done something wrong there).

Did not open issue for this beforehand but spoke to @cooperdff on Twitter about the idea and thought it was a good idea. 

Test results below:

image

### CODE REVIEW

File: nfl_data_py/__init__.py [hunk 1]
```diff
@@ -228,14 +260,22 @@ def import_weekly_data(years, columns=None, downcast=True):
...
-    if columns is None:
+    if not columns:
         columns = []
         
     # read weekly data
     url = r'https://github.com/nflverse/nflverse-data/releases/download/player_stats/player_stats_{0}.parquet'
     data = pandas.concat([pandas.read_parquet(url.format(x), engine='auto') for x in years])
 
-    if len(columns) > 0:
+    if thread_requests:
```
Comment (alecglen) [inline nfl_data_py/__init__.py:270]: Skipped over a little logic change here. As is, it always make the original sequential requests, then optionally requests them _again_ if `thread_requests=True`.

It should be something like 
```
columns = columns or None
if thread_requests:
    with ThreadPoolExecutor() as executor:
        futures = [executor.submit(_read_parquet, url.format(year), year, columns) for year in years]
        data = [future.result() for future in as_completed(futures)]
        data = sorted(data, key=lambda x: x['year'])
        data = [x['data'] for x in data]
else:
    data = [pandas.read_parquet(url.format(x), columns=columns) for x in years]
```

(bonus refactoring to get rid of the columns post-filtering just below this)

File: nfl_data_py/tests/nfl_test.py [hunk 2]
```diff
@@ -8,12 +8,22 @@ def test_is_df_with_data(self):
         s = nfl.import_pbp_data([2020])
         self.assertEqual(True, isinstance(s, pd.DataFrame))
         self.assertTrue(len(s) > 0)
+
+    def test_is_df_with_data_thread_requests(self):
+        s = nfl.import_pbp_data([2020, 2021], thread_requests=True)
+        self.assertEqual(True, isinstance(s, pd.DataFrame))
+        self.assertTrue(len(s) > 0)
```
Comment (alecglen) [inline nfl_data_py/tests/nfl_test.py:16]: Out of scope for this PR, but it reminded me that we should have a test for the caching behavior.
```suggestion
		
    # TODO: Add test(s) for cache=True
    
```

File: nfl_data_py/tests/nfl_test.py [hunk 3]
```diff
@@ -8,12 +8,22 @@ def test_is_df_with_data(self):
...
 class test_weekly(TestCase):
     def test_is_df_with_data(self):
         s = nfl.import_weekly_data([2020])
         self.assertEqual(True, isinstance(s, pd.DataFrame))
         self.assertTrue(len(s) > 0)
+
+    def test_is_df_with_data_thread_requests(self):
+        s = nfl.import_weekly_data([2020, 2021], thread_requests=True)
+        self.assertEqual(True, isinstance(s, pd.DataFrame))
+        self.assertTrue(len(s) > 0)
```
Comment (alecglen) [inline nfl_data_py/tests/nfl_test.py:27]: If you're feeling ambitious, you could assert that this method completes at least as fast as the non-threaded case 😄

**Review by alecglen — CHANGES_REQUESTED** [review]
Great idea, @bendominguez0111! Just a couple small fixes needed, plus a question on the dependency addition.

**Comment by bendominguez0111** [comment]
Cool! I will make some of these changes. I actually thought up another optimization for this but never got to it (removes the need to sort the CSVs at the end, which takes a considerable amount of time if youre pulling a lot of data) I'll add that in as well along with the changes requested

File: nfl_data_py/__init__.py [hunk 4]
```diff
@@ -44,8 +47,30 @@
...
+def _read_parquet(path, year, columns):
+    """Utility function to read parquet files in parallel
+    Args:
+        path (str): URL path to parquet file
+        year (int): year of data
+        columns (List[str]): only return these columns
+    Returns:
+        dict: year and data
+    """
+    pa = pandas.read_parquet(path, columns=columns, engine='pyarrow')
```
[the remark this answers is not in the document]
Comment (alecglen) [reply nfl_data_py/__init__.py:59]: Adding to this - I ran some performance benchmarks for the two submodules out of curiosity. They were practically identical:
image

**Comment by bendominguez0111** [comment]
Made those changes. Didn't move the `if thread_requests` block all the way down to the `# load data` comment cause it didnt exactly work, but it doesnt bypass caching now w/ this new code. Also set engine = auto instead of explicitly setting pyarrow

**Comment by bendominguez0111** [comment]
Oh, and added some additional logic compared to last time to avoid sorting years at the end. Since the HTTP requests can resolve out of order, originally there would need to be a sort at the end which took up some time. Now just creating a fixed sized list and then inserting responses into it as their threads resolve

**Review by alecglen — APPROVED** [review]
Great idea on avoiding the extra sort operation, Ben! All seems to work as intended, just had a couple more quick suggestions you can commit if you like. We'll get this merged soon.

**Comment by bendominguez0111** [comment]
Committed those suggestions. Good call, those lines were a bit cluttered

**Pull request merged**